File: backup/src/jclang/IndexAttribute.java


package jclang;

import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import jclang.structs.CXIdxAttrInfo;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class IndexAttribute {
    public enum Kind {
        UNEXPOSED,
        IB_ACTION,
        IB_OUTLET,
        IB_OUTLET_COLLECTION
    }

    private final Kind kind;
    private final Cursor cursor;
    private final IndexLocation location;

    public IndexAttribute(@NotNull CXIdxAttrInfo info) {
        this.kind = Kind.values()[info.kind];
        this.cursor = new Cursor(info.cursor);
        this.location = new IndexLocation(info.loc);
    }

    @NotNull
    /* package */ static List<IndexAttribute> createFromNative(@Nullable PointerByReference attributes, int numAttributes) {
        if (attributes == null || numAttributes == 0) {
            return Collections.emptyList();
        }
        CXIdxAttrInfo attrInfo = new CXIdxAttrInfo(attributes.getValue());
        List<IndexAttribute> result = new ArrayList<IndexAttribute>(numAttributes);
        for (int i = 0; i < numAttributes; i++) {
            // TODO: ugly hack, figure out how to make toArray() stuff work here
            CXIdxAttrInfo info = new CXIdxAttrInfo(attributes.getPointer().getPointer((long) i * attrInfo.size() / Native.POINTER_SIZE));
            result.add(new IndexAttribute(info));
        }
        return result;
    }

    @NotNull
    public Kind getKind() {
        return kind;
    }

    @NotNull
    public Cursor getCursor() {
        return cursor;
    }

    @NotNull
    public IndexLocation getLocation() {
        return location;
    }
}

File: src/main/java/klang/jvm/IndexAttribute.kt
package klang.jvm

import com.sun.jna.Native
import com.sun.jna.ptr.PointerByReference
import klang.jvm.binding.CXIdxAttrInfo

class IndexAttribute(info: CXIdxAttrInfo) {
    enum class Kind {
        UNEXPOSED,
        IB_ACTION,
        IB_OUTLET,
        IB_OUTLET_COLLECTION
    }

    val kind: Kind
    val cursor: Cursor
    val location: IndexLocation

    init {
        kind = Kind.values()[info.kind]
        cursor = Cursor(info.cursor)
        location = IndexLocation(info.loc)
    }

    companion object {
        @JvmStatic
        fun createFromNative(attributes: PointerByReference?, numAttributes: Int): List<IndexAttribute> {
            if (attributes == null || numAttributes == 0) {
                return emptyList()
            }
            val attrInfo = CXIdxAttrInfo(attributes.value)
            val result: MutableList<IndexAttribute> = ArrayList(numAttributes)
            for (i in 0 until numAttributes) {
                // TODO: ugly hack, figure out how to make toArray() stuff work here
                val info =
                    CXIdxAttrInfo(attributes.pointer.getPointer((i * attrInfo.size() / Native.POINTER_SIZE).toLong()))
                result.add(IndexAttribute(info))
            }
            return result
        }
    }
}